docs: add how-to on DinD in Kata (#3137)
kata-containers/kata-containers: docs/how-to/how-to-run-dind-with-kata.md

> # How to run Docker in Docker with Kata Containers

Reviewer: - Please add a link to this document from `docs/how-to/README.md` or the CI will fail this PR as the doc is "unreferenced".
- I would consider adding a "Note" (see https://github.com/kata-containers/kata-containers/blob/main/docs/Documentation-Requirements.md#notes) somewhere near the top of this document explaining that you can't actually run Kata 2.x with Docker to avoid misleading the reader.

Author: - [ ] Add note regarding the inability to run Kata 2.x with docker

Author: @jodh-intel do you think maybe I should rename this nested containerization or something? The whole "Docker in Docker" phrasing itself might be a little misleading in and of itself

Author: I've gone with `how-to-run-docker-with-kata.md` as the title and will alter references to DinD

> This document describes the why and how behind running Docker-in-Docker ("dind") in a kata container.

Author: - [ ] kata -> Kata on line 3

## Requirements

- A working Kata Containers installation

## Install and configure Kata Containers

> Follow the [Kata Containers installation guide](../docs/install/README.md) to Install Kata Containers on your Kubernetes cluster.

Reviewer 2: ```suggestion
Follow the [Kata Containers installation guide](/docs/install/README.md) to Install Kata Containers on your Kubernetes cluster.
```

Otherwise we link to `docs/docs/install/...`

Reviewer: Or the simplest which is `../install/README.md`. The CI checks all markdown links so it will let you know if you get it wrong :smile:

Author: Ahh thanks for this -- will put in `../install/README.md`

- [ ] Fix reference to installation guide @ `how-to-run-dind-with-kata.md#L11`

## Background

> Docker in Docker ("DinD") is the colloquial name for the ability to run `docker` from inside a container (whether provided by `docker` or otherwise) pattern for running `docker` from *inside* an existing `docker` container. You can read more about Docker-in-docker at the following links:

Reviewer: - Please fold long lines as subsequent changes make diffs hard to read.
- You might want to change some of the `docker` instances to Docker to refer to the project name, rather than the binary name (which is also slightly easier on the eye).

Author: - [ ] fold long line @ L15
- [ ] better distinguish between instances of `docker` to "Docker"

- [The original announcement of DinD](https://www.docker.com/blog/docker-can-now-run-within-docker/)

> - [`docker` image DockerHub page](https://hub.docker.com/_/docker/) (this page lists the `-dind` releases)

Reviewer 2: ```suggestion
- [`docker` image Docker Hub page](https://hub.docker.com/_/docker/) (this page lists the `-dind` releases)
```

`DockerHub` is not part of our dictionary, but `Docker Hub`seems to be fine :-)

Author: Ah I'll fix this as well -- since I'm rewriting the commit I'll do this manually rather than committing this suggestion

- [ ] `DockerHub` -> `Docker Hub` on line 18

**NOTE** While in the past kata containers did not contain the [`overlay` kernel module (aka OverlayFS)][overlayfs], it *now does* with the merging of the following pull requests:

- [PR #1384](https://github.com/kata-containers/kata-containers/pull/1384)

> - [PR #716](https://github.com/kata-containers/packaging/pull/716)).

Reviewer: It would be better to remove these links and state that Kata version x.y.z supports it.

Author: - [ ] Remove links to PRs, replace with version of Kata in which the PRs landed

> `overlayfs` is the preferred containerization storage driver for linux systems ([Docker in particular](https://docs.docker.com/storage/storagedriver/select-storage-driver)), and as such, using it is often desired over other alternatives.

Reviewer 2: ```suggestion
`OverlayFS` is the preferred containerization storage driver for Linux systems ([Docker in particular](https://docs.docker.com/storage/storagedriver/select-storage-driver)), and as such, using it is often desired over other alternatives.
```

overlayfs -> OverlayFS is a suggestion of mine as what we're adding to the dict in the PR that I mentioned is OverlayFS.

linux -> Linux is a suggestion from the spell-checker.

Author: Perfect I'll add this too:

- [ ] `overlayfs` -> `OverlayFS` on line 25
- [ ] "linux" -> "Linux" on line 25

Author: oh also here I think OverlayFS is more the project/effort name and not a binary in particular so I'll drop the backticks

> [overlayfs]: https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html

Reviewer 2: ```suggestion
[OverlayFS]: https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html
```

Author: - [ ] overlayfs -> OverlayFS on line 27

## Why DinD containers require special measures

> Running containers DinD-capable containers with Kata takes more effort than normal due to the fact that `VOLUME`s specified in `Dockerfile`s run by kata containers are given the `kataShared` mount type by default, as it is what is used for the root directory `/`:

Reviewer: Please always refer to "Kata Containers".

Author: - [ ] "kata" -> "Kata Containers" @ L31

```console
/ # mount
kataShared on / type virtiofs (rw,relatime,dax)
```

`kataShared` mount types are powered by [`virtio-fs`][virtio-fs], a marked improvement over `virtio-9p`, thanks to [PR #1016](https://github.com/kata-containers/runtime/pull/1016). While `virtio-fs` is normally an excellent choice, in the case of DinD workloads `virtio-fs` causes an issue -- [it *cannot* be used as a "upper layer" of `overlayfs` without a custom patch](http://lists.katacontainers.io/pipermail/kata-dev/2020-January/001216.html).

As `/var/lib/docker` is a `VOLUME` specified by DinD (i.e. the `docker` images tagged `*-dind`/`*-dind-rootless`), `docker` fill fail to start (or even worse, silently pick a worse storage driver like `vfs`) when started in a Kata Container. Special measures must be taken when running DinD-powered workloads in Kata Containers.

## Workarounds/Solutions

> Thanks to various community contributions (see [issue references below](#references)) the following options, with various tradeoffs have been uncovered:

Reviewer 2: ```suggestion
Thanks to various community contributions (see [issue references below](#references)) the following options, with various trade-offs have been uncovered:
```

tradeoffs -> trade-offs, in order to make the spell-checker happy.

Author: We must keep the spell-checker happy
- [ ] "tradeoffs" -> "trade-offs" on line 44

          apk add e2fsprogs &&
          truncate -s 20G /tmp/disk.img &&
          mkfs.ext4 /tmp/disk.img &&
          mount /tmp/disk.img /var/lib/docker; fi &&
        dockerd-entrypoint.sh;
      securityContext:
        privileged: true
```

Note that `loop` mounted disks are often sparse, which means they *do not* take up the full amount of space that has been provisioned. This solution seems to produce the best performance and flexibility, at the expense of increased complexity and additional required setup.

### Build a custom kernel/Wait for upstream resolution

`virtio-fs` may be supported as an upper layer for `virtio-fs` in future kernel releases, and it's possible to [modify the kernel](https://github.com/kata-containers/runtime/issues/1888#issuecomment-616872558) (in addition to applying the earlier mentioned mailing list patch) to support using `virtio-fs` as an upper. Note that if you modify your kernel and use `virtio-fs` you may require [additional changes](https://github.com/kata-containers/runtime/issues/1888#issuecomment-739057384) for decent performance and to address other issues.

## References

> The solutions proposed in this document are an amalgamation of thoughtful contributions from the kata containers community. Find links to discussion and the fruits therein below:

Reviewer: "Kata Containers".

Author: - [ ] "kata containers" -> "Kata Containers" @ L117

- [How to run Docker in Docker with Kata Containers (`runtime`#2474)](https://github.com/kata-containers/kata-containers/issues/2474)

> - [Does Kata-conatiner support aufs/overlay filesystem? (`runtime`#2493)](https://github.com/kata-containers/runtime/issues/2493)

Reviewer 2: ```suggestion
- [Does Kata-conatiner support AUFS / OverlayFS? (`runtime`#2493)](https://github.com/kata-containers/runtime/issues/2493)
```

Author: - [ ] aufs/overlay filesystem -> "AUFS / OverlayFS" on line 120
